# Consequences of emptyness

Emptyness is not the simplest of concepts for a computer.  To represent true emptyness, a comptuer needs to be doing nothing - which begs the question "how do you process emtptyness?".  The answer is that we designate special values as representing emptyness, we explore these here

## Zero is not nothing

Consider the rainfall data for Mount Boyce

In [8]:
import pandas

mtboyce = pandas.read_csv("data/rainfall/IDCJAC0009_063292_1800_Data.csv")
mtboyce[mtboyce["Rainfall amount (millimetres)"] == 0]

Product code  Bureau of Meteorology station number  Year  Month  Day  \
514     IDCJAC0009                                 63292  1995      5   30   
515     IDCJAC0009                                 63292  1995      5   31   
516     IDCJAC0009                                 63292  1995      6    1   
517     IDCJAC0009                                 63292  1995      6    2   
522     IDCJAC0009                                 63292  1995      6    7   
...            ...                                   ...   ...    ...  ...   
10609   IDCJAC0009                                 63292  2023      1   18   
10617   IDCJAC0009                                 63292  2023      1   26   
10620   IDCJAC0009                                 63292  2023      1   29   
10623   IDCJAC0009                                 63292  2023      2    1   
10624   IDCJAC0009                                 63292  2023      2    2   

       Rainfall amount (millimetres)  \
514                              0.0   
515                              0.0   
516                              0.0   
517                              0.0   
522                              0.0   
...                              ...   
10609                            0.0   
10617                            0.0   
10620                            0.0   
10623                            0.0   
10624                            0.0   

       Period over which rainfall was measured (days) Quality  
514                                               NaN       N  
515                                               NaN       N  
516                                               NaN       N  
517                                               NaN       N  
522                                               NaN       N  
...                                               ...     ...  
10609                                             1.0       N  
10617                                             1.0       N  
10620                                             1.0       N  
10623                                             1.0       N  
10624                                             1.0       N  

[4947 rows x 8 columns]

There are 531 rows where the rainfall measurements is missing, and there are 4947 rows were the rainfall measured was 0.  _These are clearly different things_.  If you were to put zero in all the empty spots, you would be suggesting an extra 531 days were it didn't rain.  There was surely rain on at least some of those days, even if it didn't get measured.

Keeping the concept of "Zero" separate from "emptyness" is very important.

The main type of emptyness to consider in pandas is `NaN` (not a number).  Where pandas expects a number, but sees nothing, it will put this special value.  `NaN`s are treated very differently to 0.  Note that `NaN` comes from the `numpy` library which pandas has been using all along without us realising and we need to import it is we want to make our own `NaN`s

In [13]:
import numpy

some_nans = pandas.Series([0, 1, 2, 3, numpy.nan, 4, 5, numpy.nan])
no_nans = pandas.Series([0, 1, 2, 3, 0, 4, 5, 0])

print(some_nans.mean())
print(no_nans.mean())

print(some_nans.sum())
print(no_nans.sum())

print("\n\n--- info ---")
print(some_nans.info())
print("\n\n--- info ---")
print(no_nans.info())

2.5
1.875
15.0
15


--- info ---
<class 'pandas.core.series.Series'>
RangeIndex: 8 entries, 0 to 7
Series name: None
Non-Null Count  Dtype  
--------------  -----  
6 non-null      float64
dtypes: float64(1)
memory usage: 196.0 bytes
None


--- info ---
<class 'pandas.core.series.Series'>
RangeIndex: 8 entries, 0 to 7
Series name: None
Non-Null Count  Dtype
--------------  -----
8 non-null      int64
dtypes: int64(1)
memory usage: 196.0 bytes
None


As you can see, you can the _aggregation_ operations ignore `NaN`s, treating them as missing instead of as 0.  Thus you need to be quite carefull with what you do to `NaN`s

# Finding NaNs

pandas has methods to help us find NaNs
  * isna()
  * isnull()

In [14]:
some_nans.isna()

0    False
1    False
2    False
3    False
4     True
5    False
6    False
7     True
dtype: bool

In [8]:
some_nans.isnull()

0    False
1    False
2    False
3    False
4     True
5    False
6    False
7     True
dtype: bool

They both do the same thing!  Why is there two names?  Ther are actually even more names!  "NaN", "na", "Null", "None" are all names you will see used for missing values.  There _are sometimes_ differences between them, but for our purposes we will treat them as synonyms.

# Exercise - copy me

Work out the two lines of pandas I used to work out how many missing values and how many zeros there were in the mtboyce rainfall column.

In [41]:
null = mtboyce["Rainfall amount (millimetres)"].isnull()
null[null == True]

# series length is 10625; there are 531 null values; total valid rows of "Rainfall amount (millimetres)" is 10,094 



0       True
1       True
2       True
3       True
4       True
        ... 
7754    True
8722    True
8723    True
9103    True
9221    True
Name: Rainfall amount (millimetres), Length: 531, dtype: bool

# Dealing with missing values

You have three options:
  * fillna
  * dropna
  * do it by hand

fillna will put some new value everywhere there is a missing value.  dropna will drop all rows that have a missing value in any of the columns.  Both of these are very blunt instruments.  We strongly recommend you do it by hand.  Make a judgement call, and write the pandas to put effect to your choice.

# Exercise - drop it like its hot

- Try out `dropna` on the Mt Boyce data frame.

- How many rows did you lose?

- Can you explain why that many were lost?

In [31]:
null = mtboyce["Rainfall amount (millimetres)"].dropna()
null
# Series length is now 10,094; 531 rows dropped because the code identified 531 null values in "Rainfall amount (millimetres)" column


null_all = mtboyce.dropna()
null_all
# Series length is now 8,658; 1,967 rows dropped because the code identified 1,967 null values across **all** columns

Product code  Bureau of Meteorology station number  Year  Month  Day  \
177     IDCJAC0009                                 63292  1994      6   27   
178     IDCJAC0009                                 63292  1994      6   28   
179     IDCJAC0009                                 63292  1994      6   29   
180     IDCJAC0009                                 63292  1994      6   30   
181     IDCJAC0009                                 63292  1994      7    1   
...            ...                                   ...   ...    ...  ...   
10620   IDCJAC0009                                 63292  2023      1   29   
10621   IDCJAC0009                                 63292  2023      1   30   
10622   IDCJAC0009                                 63292  2023      1   31   
10623   IDCJAC0009                                 63292  2023      2    1   
10624   IDCJAC0009                                 63292  2023      2    2   

       Rainfall amount (millimetres)  \
177                              0.4   
178                              5.0   
179                              0.8   
180                              0.2   
181                              0.8   
...                              ...   
10620                            0.0   
10621                            3.8   
10622                           18.8   
10623                            0.0   
10624                            0.0   

       Period over which rainfall was measured (days) Quality  
177                                               1.0       N  
178                                               1.0       N  
179                                               1.0       N  
180                                               1.0       N  
181                                               1.0       N  
...                                               ...     ...  
10620                                             1.0       N  
10621                                             1.0       N  
10622                                             1.0       N  
10623                                             1.0       N  
10624                                             1.0       N  

[8658 rows x 8 columns]

In [97]:
# fillna test 
notnull_pre = mtboyce["Rainfall amount (millimetres)"].notnull()
rainfall_mean_pre = mtboyce[notnull]["Rainfall amount (millimetres)"].mean().round(2)
fillna = mtboyce["Rainfall amount (millimetres)"].fillna(rainfall_mean_pre)
rainfall_mean_post = fillna.mean().round(2)
notnull_post = fillna.notnull()

print("pre-adjusted rainfall mean: "+str(rainfall_mean_pre)+" mm (n"+str(notnull_pre.sum())+")")
print("post-adjusted rainfall mean: "+str(rainfall_mean_post)+" mm (n"+str(notnull_post.sum())+")")


pre-adjusted rainfall mean: 2.93 mm (n10094)
post-adjusted rainfall mean: 2.93 mm (n10625)


# End of Material